sdks/python/apache_beam/io/gcp/bigtableio.py

```diff
@@ -36,43 +36,170 @@
-                         'table_id': table_id}
+                         'table_id': table_id,
+                         'filter_': filter_}
+    self.sample_row_keys = None
+    self.table = None
+    self.row_count = Metrics.counter(self.__class__.__name__, 'Row count')
+
+  def __getstate__(self):
+    return self.beam_options
+
+  def __setstate__(self, options):
+    self.beam_options = options
+    self.table = None
+    self.sample_row_keys = None
+    self.row_count = Metrics.counter(self.__class__.__name__, 'Row count')
+
+  def _get_table(self):
+    if self.table is None:
+      self.table = Client(project=self.beam_options['project_id'])\
+                      .instance(self.beam_options['instance_id'])\
+                      .table(self.beam_options['table_id'])
+    return self.table
+
+  def get_sample_row_keys(self):
+    """ Get a sample of row keys in the table.
+
+    The returned row keys will delimit contiguous sections of the table of
+    approximately equal size, which can be used to break up the data for
+    distributed tasks like mapreduces.
+    :returns: A cancel-able iterator. Can be consumed by calling ``next()``
+    			  or by casting to a :class:`list` and can be cancelled by
+    			  calling ``cancel()``.
+    """
+    if self.sample_row_keys is None:
+      self.sample_row_keys = list(self._get_table().sample_row_keys())
+    return self.sample_row_keys
+
+  def get_range_tracker(self, start_position=b'', stop_position=b''):
+    if stop_position == b'':
+      return LexicographicKeyRangeTracker(start_position)
+    else:
+      return LexicographicKeyRangeTracker(start_position, stop_position)
+
+  def estimate_size(self):
+    return list(self.get_sample_row_keys())[-1].offset_bytes
+
+  def split(self, desired_bundle_size=None, start_position=None, stop_position=None):
+    """ Splits the source into a set of bundles, using the row_set if it is set.
+
+    *** At this point, only splitting an entire table into samples based on the sample row keys is supported ***
+
+    :param desired_bundle_size: the desired size (in bytes) of the bundles returned.
+    :param start_position: if specified, the position must be used as the starting position of the first bundle.
+    :param stop_position: if specified, the position must be used as the ending position of the last bundle.
+    Returns:
+    	an iterator of objects of type 'SourceBundle' that gives information about the generated bundles.
+    """
+
+    if desired_bundle_size is not None or start_position is not None or stop_position is not None:
+      raise NotImplementedError
+
+    # TODO: Use the desired bundle size to split accordingly
+    # TODO: Allow users to provide their own row sets
+
+    sample_row_keys = list(self.get_sample_row_keys())
+    bundles = []
+    if len(sample_row_keys) > 0 and sample_row_keys[0] != b'':
+        bundles.append(SourceBundle(sample_row_keys[0].offset_bytes, self, b'', sample_row_keys[0].row_key))
```

Comment:
Please fix the formatting here... probably new lines before `SourceBundle` and the 2nd-4th varaibles)

Reply:
Done. Also added a blank line afterwards.